api/agent/drivers/docker/docker.go

```diff
@@ -269,55 +269,98 @@ func dockerMsg(derr *docker.Error) string {
 	}
-	defer cancel() // do this so that after Run exits, collect stops
-	go drv.collectStats(ctx, container, task)
+	defer cancel2()
+
+	// start the collection of statistics
+	// this will continue until ctxForStats is cancelled
+	go drv.collectStats(ctxForStats, container, task)
 
 	mwOut, mwErr := task.Logger()
 
-	waiter, err := drv.docker.AttachToContainerNonBlocking(ctx, docker.AttachToContainerOptions{
+	waiter, err := drv.docker.AttachToContainerNonBlocking(ctx2, docker.AttachToContainerOptions{
 		Container: container, OutputStream: mwOut, ErrorStream: mwErr,
 		Stream: true, Logs: true, Stdout: true, Stderr: true,
 		Stdin: true, InputStream: task.Input()})
-	if err != nil && ctx.Err() == nil {
-		// ignore if ctx has errored, rewrite status lay below
+	if err != nil && ctx2.Err() == nil {
+		// ignore if ctx2 has errored, rewrite status lay below
+		cancelCtxForStats()
 		return nil, err
 	}
 
-	err = drv.startTask(ctx, container)
-	if err != nil && ctx.Err() == nil {
+	err = drv.startTask(ctx2, container)
+	if err != nil && ctx2.Err() == nil {
 		// if there's just a timeout making the docker calls, drv.wait below will rewrite it to timeout
+		cancelCtxForStats()
 		return nil, err
 	}
 
-	return &waitResult{
-		container: container,
-		waiter:    waiter,
-		drv:       drv,
-	}, nil
+	waitResult := newWaitResult(container, waiter, drv)
+
+	go func() {
+		// wait for the container to have stopped and then cancel the collection of statistics
+		waitResult.Wait(ctx0)
+		cancelCtxForStats()
+	}()
+
+	return waitResult, nil
 }
 
-// implements drivers.WaitResult
+// waitResult implements drivers.WaitResult
+// This implementation allows multiple concurrent calls to Wait()
 type waitResult struct {
-	container string
-	waiter    docker.CloseWaiter
-	drv       *DockerDriver
+	container    string
+	waiter       docker.CloseWaiter
+	drv          *DockerDriver
+	closedSignal chan struct{} // used by the first call to Wait() to signal to subsequent calls that w.waiter is already closed
+}
+
+// newWaitResult returns a new waitResult which implements drivers.WaitResult
+// This implementation allows multiple concurrent calls to Wait()
+func newWaitResult(container string, waiter docker.CloseWaiter, drv *DockerDriver) *waitResult {
+	return &waitResult{
+		container:    container,
+		waiter:       waiter,
+		drv:          drv,
+		closedSignal: make(chan struct{}),
+	}
+
 }
 
+// waitResult implements drivers.WaitResult
+// This implementation allows multiple concurrent calls to Wait()
 func (w *waitResult) Wait(ctx context.Context) (drivers.RunResult, error) {
 	defer func() {
+		// Only one caller can call w.waiter.Close() without getting a panic.
+		// This caller will then call w.waiter.Wait() and then close closeSignal to tell the other threads that processing is finished
+		// Other callers will get a panic.
+		// They will recover() here and wait for closeSignal to be closed before returning
+		defer func() {
+			if recover() != nil {
```

Comment:
I'm hoping we can get rid of this as alluded to in previous comment.

Reply:
Agreed (as noted above)